crates/biome_js_analyze/src/semantic_analyzers/nursery/no_global_assign.rs

```diff
@@ -0,0 +1,141 @@
+use crate::globals::browser::BROWSER;
+use crate::globals::node::NODE;
+use crate::globals::runtime::{BUILTIN, ES_2021};
+
+use crate::semantic_services::Semantic;
+use biome_analyze::{context::RuleContext, declare_rule, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_js_syntax::AnyJsExpression;
+use biome_rowan::AstNode;
+
+declare_rule! {
+    /// Disallow assignments to native objects or read-only global variables.
+    ///
+    /// JavaScript environments contain numerous built-in global variables, such as 'window' in browsers and 'process' in Node.js.
+    /// Assigning values to these global variables can lead to problems, as it might overwrite essential functionality.
+    ///
+    /// Source: https://eslint.org/docs/latest/rules/no-global-assign
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// Object = null;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// window = 1;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// undefined = 1;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// /*global a:readonly*/
+    /// a = 1;
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// const a = 1;
+    /// let b = 1;
+    /// b = 100;
+    ///
+    /// onload = function() {};
+    ///
+    /// /*global a:writable*/
+    /// a = 100;
+    /// ```
+    ///
+    pub(crate) NoGlobalAssign {
+        version: "next",
+        name: "noGlobalAssign",
+        recommended: false,
+    }
+}
+
+impl Rule for NoGlobalAssign {
+    type Query = Semantic<AnyJsExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let expr = match node {
+            AnyJsExpression::JsAssignmentExpression(expr) => expr,
+            _ => return None,
+        };
+
+        let left_expr = expr.left().ok()?;
+        let right_expr = expr.right().ok()?;
+        let left = left_expr.text();
+        let right_text = right_expr.text();
+
+        let is_function_call = match right_expr {
+            AnyJsExpression::JsFunctionExpression(expr) => expr.function_token().is_ok(),
+            _ => false,
+        };
+
+        let assignment = left_expr.as_any_js_assignment()?;
+        let token = assignment
+            .as_js_identifier_assignment()?
+            .name_token()
+            .ok()?;
+        let comment = token.text();
+        let search_term = format!("global {}:readonly", left);
+
+        let global_text = if NODE.contains(&left.as_str())
+            || BROWSER.contains(&left.as_str())
+            || BUILTIN.contains(&left.as_str())
+            || ES_2021.contains(&left.as_str())
```

Comment:
These arrays are sorted. Thus, you could use `binary_search` to speed up the search.

```suggestion
        let global_text = if NODE.contains(&left.as_str())
            || BROWSER.contains(&left.as_str())
            || BUILTIN.contains(&left.as_str())
            || ES_2021.contains(&left.as_str())
```